### ISSUE TITLE
Extend Consent Mode conditions for determining Ads connection status to include a check for an Ads tag as a destination of a Google Tag

### ISSUE THREAD

**Issue by techanvil** [google/site-kit-wp#9147]
## Feature Description

For the Consent Mode feature, we have so far implemented the check for Ads being connected to look for the presence of the Ads Conversion ID or an Ads link.

We should extend these conditions to include the additional check discussed under ["when does this apply?"](https://docs.google.com/document/d/1KSTPqXNlGcN-krK5pxzdP2a-sHaVO9RYzIEU7qIF9qA/edit?tab=t.0#heading=h.t6jfalui4jsi) in the one-pager - checking for the presence of Ads tag as a destination of the connected Google Tag.

Note that we _are_ now able to use the Tag Manager API to look up a container by tag ID, so this is now technically feasible. See the [`tagId` parameter for the `containers/lookup`](https://developers.google.com/tag-platform/tag-manager/api/v2/reference/accounts/containers/lookup#parameters) endpoint and the [feature request where it was introduced](https://issuetracker.google.com/issues/330844353).

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

- Within the Consent Mode feature, the conditions for determining whether Ads is connected should include a check for the presence of an Ads tag as a destination of the connected Google Tag.
  - The Google Tag for the connected Analytics property should be checked.
  - When Tag Manager is connected the tags for the connected container should be checked.
   - Note that the case where a variable is used for the tag configuration will be handled in a separate issue: #9556
- In terms of user-visible behaviour:
  - The Setup CTA Banner should be displayed when an Ads tag is detected (while the Consent Mode `enabled` setting is `false` and the banner has not been dismissed).
  - The "Recommended" badge and the notice reading "If you have Google Ads campaigns for this site, it’s highly recommended to enable Consent mode..." on the Consent Mode section of the Settings screen should be displayed when an Ads tag is detected.
  - The [Ads users' variant](https://www.figma.com/design/mdQi9qajirgoMHKy4DiVa8/Consent-Mode?node-id=514-5375&node-type=frame&t=0Hb3Ar6KXMU4hrNJ-0) of the confirmation dialog shown when toggling the Consent Mode switch off on the Settings screen should be displayed when an Ads tag is detected.

## Implementation Brief

**Consent Mode backend updates**
* [ ] Update `includes/Core/Consent_Mode/REST_Consent_Mode_Controller.php`, creating a new REST endpoint `ads-measurement-status`. The endpoint should return an object with a single boolean property named `connected`. In this endpoint:
   * [ ] First check if the Ads module is connected using the `is_connected` method, if this is true, return `connected: true` on the endpoint immediately.
   * [ ] Next, check if Analytics is connected using the `is_connected` method. If so:
      * [ ] Check if Ads is linked to Analytics by checking the `adSenseLinked` setting, if so return `connected: true` on the endpoint immediately.
      * [ ] Check if an Ads tag is the destination of the connected Analytics property:
         * [ ] Use the Analytics Settings class with `use Google\Site_Kit\Modules\Analytics_4\Settings as Analytics_Settings;`
         * [ ] Get the Analytics settings using `$analytics_settings = ( new Analytics_Settings( $this->options ) )->get();`
         * [ ] Check `$analytics_settings['googleTagContainerDestinationIDs']` is an array using `is_array`.
         * [ ] If so, loop through each tag destination ID using `foreach`, and check if any of the IDs start with `AW-` using `substr( $analytics_setttings['googleTagContainerDestinationIDs'], 0, 3 ) === 'AW-'` and return `connected: true` at the first match.
   * [ ] Finally, check if Tag Manager is connected using the `is_connected` method, if so:
      * [ ] Use `$this->get_tagmanager_service()->accounts_containers_versions->live()` to get the live version of the configured GTM tag using the `accountID` and `internalContainerID` stored in the module settings:
      https://github.com/google/site-kit-wp/blob/165c8bb73003e7861b3dc2eead677d034def431e/includes/Modules/Tag_Manager.php#L328-L330
      * [ ] Use `array_search` to check if any of the destinations [in the tag[] key of the response](https://developers.google.com/tag-platform/tag-manager/api/v2/reference/accounts/containers/versions#resource) returned for this version are of the type `awct`, which is an "Google Ads Conversion Tracking" tag, return `connected: true` if so.
      * [ ] Use `array_filter` to find destinations [in the tag[] key of the response](https://developers.google.com/tag-platform/tag-manager/api/v2/reference/accounts/containers/versions#resource) where the type is `googtag`.
         * [ ] Get the tagId for this tag by filtering the `parameters` key for this tag, where `key` is `tagId`. This should give a tagId in the form `G-XXXXXXX`.
         * [ ] Loop through each of these tags, using the new [Tag Manager container lookup API](https://github.com/googleapis/google-api-php-client-services/blob/91f36e7dbcba491e6267c5a36152011b5ff638a5/src/TagManager/Resource/AccountsContainers.php#L121C1-L141C4), calling `$tagmanagerService->accounts_containers->lookup( array( 'tag_id' => $tag_id ) )`.
         * [ ] Use the [Tag Manager destinations API](https://github.com/googleapis/google-api-php-client-services/blob/91f36e7dbcba491e6267c5a36152011b5ff638a5/src/TagManager/Resource/AccountsContainersDestinations.php#L33C1-L48C4), passing the `accountId` and `containerId` returned above like so: `$tagmanagerService->accounts_containers_destinations->get("accounts/{account_id}/containers/{container_id}/destinations", array( 'destinationId' => 'G-XXXXXX') )` to get a list of destination tags for this container.
         * [ ] Loop through the returned destinations, check if any of the returned tags destinationIds start with the chars `AW-` using `substr( $analytics_setttings['googleTagContainerDestinationIDs'], 0, 3 ) === 'AW-'`, return `connected: true` if so.
   * [ ] If none of the above conditions are met, return `connected: false`.

**Datastore changes**
 * [ ] Update `assets/js/googlesitekit/datastore/site/consent-mode.js`, create a new fetch store called `fetchGetAdsMeasurementStatusStore` with the `baseName` `getAdsMeasurementStatus`, to retrieve the values from the new REST endpoint above, like so:
   * https://github.com/google/site-kit-wp/blob/165c8bb73003e7861b3dc2eead677d034def431e/assets/js/googlesitekit/datastore/site/consent-mode.js#L80-L90
 * [ ] Add a new `isAdsConnected` resolver (example below) which yields `fetchGetAdsMeasurementStatusStore.actions.fetchGetAdsMeasurementStatus();`
 * https://github.com/google/site-kit-wp/blob/6dc22333e29eb8595564b0574699d0d6c205ed11/assets/js/googlesitekit/datastore/user/tracking.js#L104-L110
 * [ ] Update the existing `isAdsConnected` selector to be a simple getter that returns the value of `connected` from the state set by `getAdsMeasurementStatus`.

**Reduce Tag Manager API Calls**
 * [ ] Update `assets/js/components/consent-mode/ConsentModeSetupCTAWidget.js`:
    * [ ] Move these conditions within the `useSelect` for `isAdsConnected`, returning `false` early if any other conditions that should prevent rendering are met.
    * https://github.com/google/site-kit-wp/blob/53da5b5f2855089e4c74fdde73c020aaa344750e/assets/js/components/consent-mode/ConsentModeSetupCTAWidget.js#L122-L127
    * [ ] Rename `isAdsConnected` to `shouldShowWidget` as all of the render logic is now contained in this call.

### Test Coverage

* Create thorough tests for each Ads connection condition in the new REST endpoint `tests/phpunit/integration/Core/Consent_Mode/REST_Consent_Mode_ControllerTest.php`.
* Simplify the `isAdsConnected` tests in `assets/js/googlesitekit/datastore/site/consent-mode.test.js`, as logic here as been moved to the backend.

## QA Brief

* Setup Site Kit, and on each setup test following scenarios:
  * Connect Ads module
    * Paste `await googlesitekit.data.select('core/site').isAdsConnected()` in the console, it should return `true`
  * Connect Analytics module that is connected with Ads
    * Paste above snippet, it should return `true`
  * Connect Tag manager module, and connect tag with Ads, following the screenshots/instructions from this [comment](https://github.com/google/site-kit-wp/issues/9147#issuecomment-2391835675)
    * Paste above snippet, it should return `true`
* Otherwise snippet should return false in other scenarios like: 
  * Having no Analytics, or Ads, or Tag manager active 
  * Having Analytics module active, but it is not connected with Ads, and Ads module is not active
  * Having Tag Manger active with Tags that have no containers pointing to the Ads
* Consent mode notifications should be showing as before in settings and in dashboard as setup CTA 

## Changelog entry

* Extend Consent Mode conditions logic.

**Comment by eugene-manuilov** [comment]
AC ✔️

**Comment by benbowler** [comment]
Hey @techanvil, I had a deep dive into the Tag Manager docs and API and it appears I can do what is required here using existing APIs. Did I miss a way that GTM can be configured where a tags/containers can be nested in a way that requires us to look up containers to check for Ads linking? If so can you share an example of how to set up tag manager to reach this case and I will update the IB.

**Comment by techanvil** [comment]
Hey @benbowler, thanks for looking into this. The trickier case that you'll need to update the IB for is when the GTM container points to a Google tag, which in turn points to an Ads tag as one of its destinations.

As a prerequisite you'll need a Google tag which points to an Ads tag. You can configure this via the Ads UI:

![image](https://github.com/user-attachments/assets/8e467333-f868-42ee-a946-b078595e7dcc)

Or by editing a tag via the entry point on https://tagmanager.google.com/#/home#tags, updating it to point to an Ads tag as a destination:

![image](https://github.com/user-attachments/assets/c2bb9bc4-17c0-47f6-96d1-a360e92dda66)

Then, when creating your container in Tag Manager, set up a tag using the Google Tag type:

![image](https://github.com/user-attachments/assets/f700517b-f004-4d57-846b-573c12defa81)

Specify the Google tag ID (it may have multiple IDs, you can choose any of them) in the Tag Configuration (or for a more advanced use case you can specify this indirectly via a variable):

![image](https://github.com/user-attachments/assets/5914ec4f-79f4-4805-b120-5c3aa3e972c7)

When you retrieve the live container version from the API you'll see the tag entry has the type `googtag`.

![image](https://github.com/user-attachments/assets/ca0ea723-19d3-4483-a6e4-d1c42c3c13af)

You'll need to use the `tagId` parameter to `containers/lookup` as mentioned in the Feature Description to retrieve the container for the Google tag, noting its account ID and container ID. Note that you can't rely on the list of `tagIds` here to determine whether the tag points to an Ads tag, as these will not necessarily map to destination IDs.

![image](https://github.com/user-attachments/assets/fd398177-d5f6-43b3-9660-fb836f81a9b9)

Then, use the `destinations` endpoint to retrieve the list of destinations for the Google tag. These can be examined to determine whether there's an Ads tag as a destination by checking for a `destinationId` with the prefix `AW-`:

![image](https://github.com/user-attachments/assets/bf25e4b1-b3e4-4b23-b9cd-3f3c92aad93b)

It's not the most straightforward one and I hope that helps, please drop me a line if you have any further questions!

**Comment by techanvil** [comment]
Hi @benbowler, and apologies, while reviewing the updated IB here I realised that I'd missed a point from the AC, which I've now added. The IB will need an update to accommodate this change:

>   - The [Ads users' variant](https://www.figma.com/design/mdQi9qajirgoMHKy4DiVa8/Consent-Mode?node-id=514-5375&node-type=frame&t=0Hb3Ar6KXMU4hrNJ-0) of the confirmation dialog shown when toggling the Consent Mode switch off on the Settings screen should be displayed when an Ads tag is detected.

Additionally:

- It looks like the intention of the specced `recommend-consent-mode` endpoint is to provide the same functionality as the existing `isAdsConnected()` selector, augmented to check the connected Tag Manager tag when present. However, as specced the endpoint is missing the [check for an Ads tag as a destination of the connected Analytics property](https://github.com/google/site-kit-wp/blob/53da5b5f2855089e4c74fdde73c020aaa344750e/assets/js/googlesitekit/datastore/site/consent-mode.js#L369-L376). Assuming we continue with this endpoint, this logic would need to be included.
- I'm not averse to creating a new endpoint to encapsulate all the logic on the backend. We could alternatively take the approach of adding an endpoint _just_ for the Tag Manager checks, and add a check for this to the existing `isAdsConnected()` selector.
- If we do fully migrate the logic to the new endpoint maybe the endpoint (and corresponding datastore functions) should be called `is-ads-connected`, `ads-connection-status`, or something along those lines, as we might want to use the same check outside of the context of recommending Consent Mode.
- All of the existing use of the `isAdsConnected` value should be replaced by the new version. I'd suggest we simply update the current selector rather than introduce a new one.
- We are adding potentially quite a few additional Tag Manager API calls to every dashboard page load. To mitigate this we should update the `useSelect()` in `ConsentModeSetupCTAWidget` that retrieves the `isAdsConnected` value (or `recommendConsentMode` if we still go that route) so that it only calls the selector if the other conditions that would prevent the CTA from displaying are all true:

https://github.com/google/site-kit-wp/blob/53da5b5f2855089e4c74fdde73c020aaa344750e/assets/js/components/consent-mode/ConsentModeSetupCTAWidget.js#L77-L79

https://github.com/google/site-kit-wp/blob/53da5b5f2855089e4c74fdde73c020aaa344750e/assets/js/components/consent-mode/ConsentModeSetupCTAWidget.js#L122-L127

- We might want to iterate on this in a future issue to cache the value and only update it say once per hour.
- There is a less common case we should handle where the tag ID is specified via a variable in the Tag Manager container configuration, which I alluded to this in my comment above. We can tackle this in a separate issue - I've created an issue for it and added a mention of this to the AC as well to make it clear.

**Comment by benbowler** [comment]
Thanks @techanvil, I'll review and update the IB.

**Comment by techanvil** [comment]
Thanks @benbowler! It's looking good. A few last points:

- On reflection I think `ads-measurement-status` would be preferable for a REST endpoint name as it's more indicative of a "resource", we can return an object with a `connected` boolean property. This would be along similar lines to the existing [`tracking` endpoint](https://github.com/google/site-kit-wp/blob/4d4224b95d6aedc3d86048ab8bd4ba77053a39ea/includes/Core/Tracking/REST_Tracking_Consent_Controller.php#L96-L142).
- We should check the connected Analytics property before checking Tag Manager as this can allow us to return early without any additional API requests.
- We still need a simple `isAdsConnected` selector to go with the resolver.

In order to help move this along I've updated the IB accordingly myself. Please take a look at my changes, if you're happy with them feel free to move this to the EB.

**Comment by benbowler** [comment]
Thanks Tom, I have reviewed this and moving it to EB.

**Comment by zutigrm** [comment]
The current `apiclient` library we are using includes `google/apiclient-services` `v0.355.0` which does not have updated Tag manager API, so `'lookup' => ['path' => 'tagmanager/v2/accounts/containers:lookup', 'httpMethod' => 'GET', 'parameters' => ['destinationId' => ['location' => 'query', 'type' => 'string']]]` lookup from this IB point:

> Loop through each of these tags, using the new [Tag Manager container lookup API](https://github.com/googleapis/google-api-php-client-services/blob/91f36e7dbcba491e6267c5a36152011b5ff638a5/src/TagManager/Resource/AccountsContainers.php#L121C1-L141C4), calling $tagmanagerService->accounts_containers->lookup( array( 'tag_id' => $tag_id ) )

Does not work, as tag ID is not supported as parameter in the package we have, hence throwing fatal error. 

Updating dependencies under `google/apiclient` is at that limit of services version, and we cannot update client version due to the minimum php version constrains (any version over current one requires php 8).

Some of the options we have:
* Apply all logic up to this point - meaning all the checks switched to the backend, except for additional lookup inside the `googtag` Tag, and file a follow up to extend that part once we switch to the generated third party instead of require from external source
* Move issue back to backlog until it is unblocked, switching to generated services dependencies, etc, and finish everything afterwards

WDYT? @techanvil @aaemnnosttv

**Comment by techanvil** [comment]
Hey @zutigrm! Thanks for the update, and sorry that we missed the lack of support for the `tag_id` parameter during definition.

I think your first suggestion makes the most sense - do as much as we can now, and file a followup issue to update the service library and implement the remaining requirements. It will still be an incremental improvement that would be better to deliver sooner rather than later, plus it will result in two smaller issues/PRs rather than one larger one, so it will be easier to review.

**Comment by benbowler** [comment]
@zutigrm did you create the follow up issue you mentioned, otherwise I can create it so it gets tracked.

**Comment by zutigrm** [comment]
Added a follow up [issue](https://github.com/google/site-kit-wp/issues/9770)

**Comment by wpdarren** [comment]
### QA Update: ✅ 

**Verified:**

- I setup Site Kit, and on each setup test following scenarios:
  - Connect Ads module - it returns `true`
  - Connect Analytics module that is connected with Ads - it returns `true`
  - Connect Tag manager module, and connect tag with Ads, - it returns `true`

- Having no Analytics, or Ads, or Tag manager active - it returns `false`
- Having Analytics module active, but it is not connected with Ads, and Ads module is not active  - it returns `false`
- Having Tag Manger active with Tags that have no containers pointing to the Ads  - it returns `false`

- Consent mode notifications shows as before in settings and in dashboard as setup CTA

<details>
 <summary>Screenshots</summary>

![Image](https://github.com/user-attachments/assets/7e929972-013d-40ad-99d8-20d44ae6f952)
![Image](https://github.com/user-attachments/assets/ed46be67-bf2d-4c56-8eb4-4974da1378bf)
![Image](https://github.com/user-attachments/assets/b68dac35-e520-4218-b803-0593a5b94efa)
![Image](https://github.com/user-attachments/assets/e2353236-b878-422b-b0c4-cfec36f0461d)
![Image](https://github.com/user-attachments/assets/2b35b145-5d77-4775-921c-0ee6611f6968)
![Image](https://github.com/user-attachments/assets/5f17c461-05f3-4497-b704-ccacdef26030)
![Image](https://github.com/user-attachments/assets/59248c1c-fd88-4f53-943c-97053b68513e)
![Image](https://github.com/user-attachments/assets/311c81d5-a7c6-441e-891c-e02bb1ba7650)

</details>
